### DISCUSSION TITLE
我想咨询一下，我能把准备就绪的videoStream推送到服务器吗？

### DISCUSSION THREAD

**Discussion by yaomoshen520** [yume-chan/ya-webadb#464] [Ideas]
![QQ截图20221112113459](https://user-images.githubusercontent.com/95835643/201454497-94b4426c-ea6d-4613-a26e-aca016e272c4.jpg)
问题是这样的，在创建完成client实例后，我是否可以通过 client.videoStream这个方法把流通过WebSockets，把这个流推送到服务器，进行远程控制和现实呢？如果可以应该是怎么推送，

**Comment by yume-chan** [comment]
Interesting usage. You can serialize the `packet` using any method you want, just notice that it may contain a `Uint8Array` that can't be serialized as JSON effeciently.

For example, you can do this:

```diff
+            // Say you already has a WebSocket connected
+            const ws: WebSocket = new WebSocket('ws://localhost:8080');
+
             client.videoStream
                 .pipeThrough(new InspectStream(action((packet: ScrcpyVideoStreamPacket) => {
                     if (packet.type === 'configuration') {
@@ -832,10 +835,18 @@ class ScrcpyPageState {
 
                         this.width = croppedWidth;
                         this.height = croppedHeight;
+
+                        ws.send(JSON.stringify(packet));
+                    } else {
+                        ws.send(JSON.stringify({ ...packet, data: undefined }));
+                        ws.send(packet.data);
                     }
                 })))
                 .pipeTo(decoder.writable)
-                .catch(() => { });
+                .catch(() => { })
+                .finally(() => {
+                    ws.close();
+                });
```

**Reply by yaomoshen520** [reply]
我尝试了一下，推送成功后是一个对象，我之前的想法是推送一个连续的字节流，然后在服务器中转发，在使用这个流，实现远程控制和显示，
![QQ截图20221114164026](https://user-images.githubusercontent.com/95835643/201614165-9d7fa111-575e-46d6-bf84-0ce7d8a89297.jpg)
这个不是实际需要的流

**Reply by yume-chan** [reply]
You only need the data, the format doesn't matter. You can serialize packets into any binary format (using msgpack, protobuf, or even my `@yume-chan/struct` package) if you want. Although I don't recommend sending raw Scrcpy stream data, because its structure depends on server version and arguments (so you must also send those to the other side for correct parsing).

**Reply by yaomoshen520** [reply]
非常感谢您能在忙碌的工作中，抽出时间回复；我对web Stream API 这个没有太多的了解，我之前就看到您的这个开源项目，因为工作的原因，没能抽出太多的时间了解，最近有空闲时间，所以研究了一下您的这个开源项目，我主要的工作是做后端数据库管理的。对于您之前的回复，我也做了尝试，我发现拿到的是一个对象，所以有点不明白，我的想法其实很简单， 我并不想尝重新构建底层数据包的封装，因为这个是一个很繁琐的过程，我有看到您的项目已中scrcpy.tsx中已经构建好了 client 这个实例，所以我当时有个想法，通过websocket 把 流推送到服务器 然后在服务器转发 到 前端展示. 附属一张,我的想法图.



![QQ截图20221110154744](https://user-images.githubusercontent.com/95835643/201883655-97f9d49e-4be1-4a19-ba07-468db3948302.jpg)

**Reply by yume-chan** [reply]
Your diagram is correct, and of course you can directly forward raw Scrcpy packets if you don't use `AdbScrcpyClient` at near-device side (the side you connect to your device).

However, as I mentioned above, the structure of Scrcpy packets depend on server version and arguments used, it might be simpler to normalize those packets, than duplicating or sending options around. `AdbScrcpyClient` is designed to do that. Using it also requires you parse serialize control messages and call the corresponding method on `client.controlMessageSerializer` at near-device side, but again, control messages are also version dependent so more complex to use directly.

With this approach, only the near-device side needs to depend on `@yume-chan/scrcpy` package, the server side and control side only operate on your custom protocol. IMO this makes the whole app much cleaner.
